// github-actions-linter/services/renaming.js
"use strict";
/*!
 * Copyright 2019 Omar Tawfik. Please see LICENSE file at the root of this repository.
 */
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
var vscode_languageserver_1 = require("vscode-languageserver");
var cache_1 = require("../util/cache");
var RenamingService = /** @class */ (function () {
    function RenamingService() {
    }
    RenamingService.prototype.fillCapabilities = function (capabilities) {
        capabilities.renameProvider = {
            prepareProvider: true,
        };
    };
    RenamingService.prototype.activate = function (connection, documents) {
        connection.onPrepareRename(function (params) {
            var uri = params.textDocument.uri;
            var compilation = cache_1.accessCache(documents, uri);
            var target = compilation.getTargetAt(params.position);
            if (!target) {
                return undefined;
            }
            return {
                placeholder: target.name,
                range: target.range,
            };
        });
        connection.onRenameRequest(function (params) {
            var _a;
            var uri = params.textDocument.uri;
            var compilation = cache_1.accessCache(documents, uri);
            var target = compilation.getTargetAt(params.position);
            if (!target) {
                return undefined;
            }
            var action = compilation.actions.get(target.name);
            if (!action) {
                return undefined;
            }
            return {
                changes: (_a = {},
                    _a[uri] = __spread([action.range], action.references).map(function (range) {
                        var start = range.start, end = range.end;
                        return vscode_languageserver_1.TextEdit.replace(vscode_languageserver_1.Range.create(start.line, start.character + 1, end.line, end.character - 1), params.newName);
                    }),
                    _a),
            };
        });
    };
    return RenamingService;
}());
exports.RenamingService = RenamingService;
