// github-actions-linter/scanning/tokens.js
"use strict";
/*!
 * Copyright 2019 Omar Tawfik. Please see LICENSE file at the root of this repository.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var TokenKind;
(function (TokenKind) {
    // Top-level keywords
    TokenKind[TokenKind["VersionKeyword"] = 0] = "VersionKeyword";
    TokenKind[TokenKind["WorkflowKeyword"] = 1] = "WorkflowKeyword";
    TokenKind[TokenKind["ActionKeyword"] = 2] = "ActionKeyword";
    // Bottom-level keywords
    TokenKind[TokenKind["OnKeyword"] = 3] = "OnKeyword";
    TokenKind[TokenKind["ResolvesKeyword"] = 4] = "ResolvesKeyword";
    TokenKind[TokenKind["UsesKeyword"] = 5] = "UsesKeyword";
    TokenKind[TokenKind["NeedsKeyword"] = 6] = "NeedsKeyword";
    TokenKind[TokenKind["RunsKeyword"] = 7] = "RunsKeyword";
    TokenKind[TokenKind["ArgsKeyword"] = 8] = "ArgsKeyword";
    TokenKind[TokenKind["EnvKeyword"] = 9] = "EnvKeyword";
    TokenKind[TokenKind["SecretsKeyword"] = 10] = "SecretsKeyword";
    // Punctuation
    TokenKind[TokenKind["Equal"] = 11] = "Equal";
    TokenKind[TokenKind["Comma"] = 12] = "Comma";
    // Brackets
    TokenKind[TokenKind["LeftCurlyBracket"] = 13] = "LeftCurlyBracket";
    TokenKind[TokenKind["RightCurlyBracket"] = 14] = "RightCurlyBracket";
    TokenKind[TokenKind["LeftSquareBracket"] = 15] = "LeftSquareBracket";
    TokenKind[TokenKind["RightSquareBracket"] = 16] = "RightSquareBracket";
    // Misc
    TokenKind[TokenKind["Identifier"] = 17] = "Identifier";
    TokenKind[TokenKind["Comment"] = 18] = "Comment";
    // Literals
    TokenKind[TokenKind["IntegerLiteral"] = 19] = "IntegerLiteral";
    TokenKind[TokenKind["StringLiteral"] = 20] = "StringLiteral";
    // Generated
    TokenKind[TokenKind["Missing"] = 21] = "Missing";
    TokenKind[TokenKind["Unrecognized"] = 22] = "Unrecognized";
})(TokenKind = exports.TokenKind || (exports.TokenKind = {}));
function getTokenDescription(kind) {
    switch (kind) {
        case TokenKind.VersionKeyword:
            return "version";
        case TokenKind.WorkflowKeyword:
            return "workflow";
        case TokenKind.ActionKeyword:
            return "action";
        case TokenKind.OnKeyword:
            return "on";
        case TokenKind.ResolvesKeyword:
            return "resolves";
        case TokenKind.UsesKeyword:
            return "uses";
        case TokenKind.NeedsKeyword:
            return "needs";
        case TokenKind.RunsKeyword:
            return "runs";
        case TokenKind.ArgsKeyword:
            return "args";
        case TokenKind.EnvKeyword:
            return "env";
        case TokenKind.SecretsKeyword:
            return "secrets";
        case TokenKind.Equal:
            return "=";
        case TokenKind.Comma:
            return ",";
        case TokenKind.LeftCurlyBracket:
            return "{";
        case TokenKind.RightCurlyBracket:
            return "}";
        case TokenKind.LeftSquareBracket:
            return "[";
        case TokenKind.RightSquareBracket:
            return "]";
        case TokenKind.Identifier:
            return "identifier";
        case TokenKind.Comment:
            return "comment";
        case TokenKind.IntegerLiteral:
            return "integer";
        case TokenKind.StringLiteral:
            return "string";
        case TokenKind.Missing:
            return "missing";
        case TokenKind.Unrecognized:
            return "unrecognized";
    }
}
exports.getTokenDescription = getTokenDescription;
