// github-actions-linter/util/constants.js
"use strict";
/*!
 * Copyright 2019 Omar Tawfik. Please see LICENSE file at the root of this repository.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MAXIMUM_SUPPORTED_VERSION = 0;
exports.MAXIMUM_SUPPORTED_SECRETS = 100;
exports.MAXIMUM_SUPPORTED_ACTIONS = 100;
exports.LANGUAGE_NAME = "github-actions";
var USES_REGEX;
(function (USES_REGEX) {
    var ALPHA_NUM = "[a-zA-Z0-9]";
    var ALPHA_NUM_DASH = "[a-zA-Z0-9-]";
    var DOCKER_HOST_COMPONENT = "(" + ALPHA_NUM + "|(" + ALPHA_NUM + ALPHA_NUM_DASH + "*" + ALPHA_NUM + "))";
    var DOCKER_REGISTRY = "(" + DOCKER_HOST_COMPONENT + "(\\." + DOCKER_HOST_COMPONENT + ")*(:[0-9]+)?\\/)";
    var DOCKER_PATH_COMPONENT = "(" + ALPHA_NUM + "+([._-]" + ALPHA_NUM + "+)*)";
    var DOCKER_TAG = "(:[a-zA-Z0-9_]+)";
    var DOCKER_DIGEST_ALGORITHM = "[A-Za-z]" + ALPHA_NUM + "*";
    var DOCKER_DIGEST = "(@" + DOCKER_DIGEST_ALGORITHM + "([+.-_]" + DOCKER_DIGEST_ALGORITHM + ")*:[a-fA-F0-9]+)";
    var DOCKER_USES = "docker:\\/\\/" + DOCKER_REGISTRY + "?" + DOCKER_PATH_COMPONENT + "(\\/" + DOCKER_PATH_COMPONENT + ")*(" + DOCKER_TAG + "|" + DOCKER_DIGEST + ")?";
    var LOCAL_USES = "\\.\\/.*";
    var REMOTE_OWNER = ALPHA_NUM + "+(" + ALPHA_NUM_DASH + ALPHA_NUM + "+)*";
    var REMOTE_PATH = "(\\/[a-zA-Z0-9-_.]+)";
    var REMOTE_REF = "@.+";
    var REMOTE_USES = "" + REMOTE_OWNER + REMOTE_PATH + "+" + REMOTE_REF;
    var COMBINED = new RegExp("^(" + DOCKER_USES + ")|(" + LOCAL_USES + ")|(" + REMOTE_USES + ")$");
    function test(value) {
        return COMBINED.test(value);
    }
    USES_REGEX.test = test;
})(USES_REGEX = exports.USES_REGEX || (exports.USES_REGEX = {}));
